Store GC identify action's results into a table

ajantha-bhat opened the pull request:
Currently, output of `GCImpl#identifyExpiredContents` is stored in an in memory map by merging the results from each spark task.

The results can be stored in a table instead, to run clean up afterwards or to be used by consumer of a different process. 

With this PR, Spark is directly writing the task output into the iceberg table in a configured reference where each task's output will be mapped to multiple rows in a table (one row per content id).

`GCImpl#identifyExpiredContents` now returns the run id of the completed task. Using which user can query the output table. Also added an interface to fetch the latest completed run id.

All old gc test cases now query the output table to collect the results and validates it.
Added a new test case to test in-progress and marker row write to the table.

Also removed `Serializable` from `Content` along with removing intermediate classes (`contentValues`,`IdentifiedResult`)that was used for storing output in-memory map.

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/GCParams.java:
```diff
@@ -70,6 +70,12 @@ default double getBloomFilterFpp() {
+  String getNessieCatalogName();
+
+  /** Output table name to be used for storing the results in default reference. */
+  String getOutputTableIdentifier();
```
using the default reference, because if we open a configuration for reference name,

a) what if the reference not exist ?
b) If need to implicitly create reference, what is the the source reference and its hash?
c) Using  SQL for query and insert. So, use reference sql need to be used. Which requires dependency on nessie sql extensions. What if user forgot to configure it ?

snazy replied:
So every little activity causes unintended conflicts on the "main" branch and "pollute" the commit log of the "main" branch.
I don't really like that approach, because it bugs users with maintenance operations that they haven't actively triggered and very likely cause a lot questions about what that is.

ajantha-bhat replied:
hmmm, good point👍🏻.
I thought about conflict before, but as in nessie conflict is per content, I thought conflict will not happen if each time GC creates a new table and other ref may not use gc_table. (but that is a wrong assumption, other ref can have this gc_table name)

I will expose a mandatory config for refName for GC, 
a) If reference does not exist, reference is created pointing to NO_ANCESTOR hash.
b) If reference exist, then use that reference.
c) SQL syntax need ref, so mandatory dependency on NessieSparkSessionExtensions

ajantha-bhat commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGCTest.java:
```diff
@@ -25,7 +25,7 @@
 import org.projectnessie.model.Branch;
 
-public abstract class AbstractRestGCTest extends AbstractRestGC {
+public abstract class AbstractRestGCTest extends AbstractRestGCRepoTest {
```
This is just an extension to run all the test cases from single consumer class, similar to `AbstractRestTest` hierarchy.

snazy commented on servers/jax-rs-tests/pom.xml:
```diff
@@ -124,6 +124,11 @@
+    <dependency>
+      <groupId>org.apache.iceberg</groupId>
+      <artifactId>iceberg-spark3-runtime</artifactId>
+      <scope>provided</scope>
```
Why is this a non-test dependency - i.e. Iceberg leaking to other modules.

ajantha-bhat replied:
for spark configuration of `spark.sql.catalog.nessie.catalog-impl`, I have used `NessieCatalog.class.getName()`.
Hence it couldn't be a test dependency. 

But now I changed conf to use `org.apache.iceberg.nessie.NessieCatalog`.
**Hence changed now to test dependency.**

snazy reviewed (changes requested):
Looks like this PR needs way more testing, espscially tests that run w/ more than one partition. Handling of the "GC-successfully-finished" marker row is not properly tested.

ajantha-bhat commented:
> Looks like this PR needs way more testing, espscially tests that run w/ more than one partition. Handling of the "GC-successfully-finished" marker row is not properly tested.

a. All previous gc testcase now writes to spark table of iceberg format directly with successful marker row. Hence all those success scenarios are tested.
b. Added new testcase for single and multiple partitions which tests the repo reading and writing with and without marker row(with a stubbed GC result rows)

Please have a look at it again.   Thanks.

ajantha-bhat commented:
Build failed now for maven connection time out. Should not be because of my changes. Looking into it.

> Error:  Failed to execute goal on project nessie-perftest-gatling: Could not resolve dependencies for project org.projectnessie:nessie-perftest-gatling:jar:0.21.3-SNAPSHOT: Could not transfer artifact io.gatling:gatling-commons:jar:3.7.6 from/to central (https://repo.maven.apache.org/maven2): transfer failed for https://repo.maven.apache.org/maven2/io/gatling/gatling-commons/3.7.6/gatling-commons-3.7.6.jar: Connect to repo.maven.apache.org:443 [repo.maven.apache.org/146.75.28.215] failed: Connection timed out (Connection timed out) -> [Help 1]

ajantha-bhat commented:
@snazy : I have made a configurable reference for GC output table and pushed as a separate commit for easy review. Please have a look.

tomekl007 commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGCRepoTest.java:
```diff
@@ -0,0 +1,192 @@
+      // should collect 5 rows.
+      assertThat(identifiedResult.getContentValuesForReference(refName).size()).isEqualTo(5);
+
+      sparkSession.sql(String.format("DROP TABLE %s", catalogAndIdentifierWithReference));
```
we should use before/after or move it to finally. It maybe not dropped in case there will be an exception before (same for other tests)

ajantha-bhat replied:
As it was a test code and test was expected to pass, I have added like that.

Moved to finally block now.

tomekl007 commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/GCImpl.java:
```diff
@@ -189,4 +192,17 @@ private static Instant getInstantFromMicros(Long microsSinceEpoch) {
+    } catch (NessieNotFoundException e) {
+      // create a gc reference pointing to NO_ANCESTOR hash.
+      try {
+        api.createReference().reference(Branch.of(refName, null)).create();
```
does `null == NO_ANCESTOR`?

tomekl007 commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/GCParams.java:
```diff
@@ -70,6 +70,20 @@ default double getBloomFilterFpp() {
+   * <p>If the reference doesn't exist for this reference name, reference with this name pointing to
+   * beginning of time (aka NO_ANCESTOR hash) will be created.
+   */
+  String getOutputTableRefName();
```
Not strictly related to this PR, but how do we handle weird table names and escaping stuff? Shouldn't we have a dedicated `TableRefName` type instead of raw `String`?

ajantha-bhat replied:
Here, raw string is fine as it is just the reference name.

But internally for DDL and DML, we need to append  `@ref` and escape characters to the table name to tell SQL parser that it is associated with some reference. 
I just discovered that we have `TableReference` class, I will try to use it internally for DDL and DML

tomekl007 commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java:
```diff
@@ -0,0 +1,322 @@
+    // collect marker row for the given run-id
+    String checkMarkerRow =
+        String.format(
+            "SELECT %s FROM %s WHERE %s = '%s' AND %s = '%s'",
+            COL_GC_RUN_ID,
+            withRefName(catalogAndTable, gcRefName),
+            COL_GC_RUN_ID,
+            runId,
+            COL_TYPE,
+            TYPE_GC_MARKER);
```
maybe extract to helper method? It is used in multiple places

ajantha-bhat replied:
sorry, where is it repeated? I think arguments are different.

rymurr replied:
I think something like the helper method in icebergs tests which makes it easier to read. Something like

`Dataset<Row> sql(sqlStatement, Object... args)`

rymurr commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/DistributedIdentifyContents.java:
```diff
@@ -66,21 +69,26 @@ public DistributedIdentifyContents(SparkSession session, GCParams gcParams) {
 
     IdentifyContentsPerExecutor executor = new IdentifyContentsPerExecutor(gcParams);
-    List<IdentifiedResult> results =
+    JavaRDD<IdentifiedResult> resultRDD =
```
why are we using a RDD instead of a dataframe. This low level API should be avoided

ajantha-bhat replied:
When I tried using DataFrame instead of RDD, DataFrame is doing codeGen with Nessie's `Reference` interface. As `Reference` doesn't have public constructor, code gen failed. If I add a public constructor with immutables, Immutables is failing to build the class saying it cannot apply validation. Need to explore further. Instead of helping Dataframe is causing problem.

rymurr commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java:
```diff
@@ -0,0 +1,322 @@
+  // and the results for that run id is consumable.
+  private static final String TYPE_GC_MARKER = "GC_MARK";
+
+  private static final String SCHEMA_DDL =
```
I think calling create from the iceberg apis or from the dataframe api would be easier. We can define the schema as avro instead of as a long and hard to read string

rymurr commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java:
```diff
@@ -0,0 +1,322 @@
+   */
+  public Dataset<Row> collectExpiredContentsAsDataSet(String runId) {
+    // collect marker row for the given run-id
+    String checkMarkerRow =
```
can we add a debug line to log this sql statement. Its hard to read like this.

rymurr commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java:
```diff
@@ -0,0 +1,322 @@
+            COL_TYPE,
+            TYPE_GC_MARKER);
+
+    List<Row> markerRows = sparkSession.sql(checkMarkerRow).collectAsList();
```
why turn into a list and back into a dataframe?

rymurr commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java:
```diff
@@ -0,0 +1,322 @@
+      return sparkSession.createDataFrame(Collections.emptyList(), schema);
+    }
+
+    String expiredContentIdsQuery =
```
same comment about log

rymurr commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java:
```diff
@@ -0,0 +1,322 @@
+   */
+  public String getLatestCompletedRunID() {
+    // collect marker row for the last written run id
+    String checkMarkerRow =
```
why not do a groupby and max? Prob faster and better on memory than a sort

ajantha-bhat replied:
If there is an in-progress entry (without marker row), max(timestamp) will select the timestamp of in-progress entry and I may not find marker row for that timestamp and query will return empty results.

So, I think current SQL is safe.

rymurr commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java:
```diff
@@ -0,0 +1,322 @@
+   * @param runId run id of current identify task.
+   * @param startedAt task start time.
+   */
+  void writeToOutputTable(JavaRDD<IdentifiedResult> resultRDD, String runId, Timestamp startedAt) {
```
i really dont like the RDD (sorry for hte 2nd comment) I would much rather we do this via normal spark.write tooling,

rymurr commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java:
```diff
@@ -0,0 +1,322 @@
+      Iterator<IdentifiedResult> result, String runID, Timestamp startedAt) {
+    List<Row> rows = new ArrayList<>();
+    ObjectMapper objectMapper = new ObjectMapper();
+    while (result.hasNext()) {
```
what is this doing? why not use a stream or something?

rymurr commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java:
```diff
@@ -0,0 +1,322 @@
+  private static void addRowToIdentifiedResults(
+      IdentifiedResult identifiedResult, ObjectMapper objectMapper, Row row) {
+    List<Object> contentsInRow = row.getList(1);
+    List<Content> contents =
```
why materialize teh List just to do a foreach on it?

ajantha-bhat replied:
yeah, no need.

rymurr commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java:
```diff
@@ -0,0 +1,322 @@
+
+  private static String withRefName(String catalogAndTable, String refName) {
+    int tableNameIndex = catalogAndTable.lastIndexOf(".") + 1;
+    return catalogAndTable.substring(0, tableNameIndex)
```
why not set the context via teh spark session or store the underlying as member variables? I don't know why we are doing string manipultion here. And I don't like explicitly relying on the `@` syntax

ajantha-bhat replied:
Don't want to affect other concurrent operation on this spark context by setting a `ref`, currently directly setting for each table ops using the `@` syntax.

I will make it a member variable and use a `TableReference` object, that internally takes care about string manipulation.

rymurr commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java:
```diff
@@ -0,0 +1,322 @@
+    JavaRDD<Row> rowJavaRDD =
+        resultRDD.mapPartitions(result -> getContentRows(result, runId, startedAt));
+    try {
+      sparkSession
```
what is the difference between a marker row and a normal row?

ajantha-bhat replied:
for a given content id, there can be more than one normal row (one row per reference). Consumer should not read this results from this table until the marker row is present for this content id.
**marker row is to avoid reading the in-progress writes.**

rymurr reviewed (changes requested):
the broad idea looks good to me but I see extra complexity that I think we can remove.

rymurr commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java:
```diff
@@ -0,0 +1,322 @@
+          + " string COMMENT 'Nessie Content.id',\n"
+          + "  "
+          + COL_EXPIRED_CONTENTS
+          + " array<String> COMMENT 'List of expired contents (as a json serialized string), can be empty',\n"
```
flatten this please. 1 row per expeired contents. Its hard to work w/ complex types

rymurr commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java:
```diff
@@ -0,0 +1,322 @@
+      "CREATE TABLE IF NOT EXISTS %s (\n"
+          + "  "
+          + COL_TYPE
+          + " string COMMENT 'Type of information this row represents, marker row or content row',\n"
```
I don't understand the marker row? can you explain?

ajantha-bhat replied:
as explained above,

> for a given content id, there can be more than one normal row (one row per reference). Consumer should not read this results from this table until the marker row is present for this content id.
marker row is to avoid reading the in-progress writes.

rymurr commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java:
```diff
@@ -0,0 +1,322 @@
+          + ")\n"
+          + "USING iceberg\n"
+          + "PARTITIONED BY ("
+          + COL_GC_RUN_START
```
I dont think we want to partition by a timestamp, rather by run id. Personally i wouldn't partition at all to start. Premature optimisation

ajantha-bhat commented:
@rymurr and @snazy : 
I have removed RDD and made it dataframe based.

Sample output from table:
<img width="1566" alt="image" src="https://user-images.githubusercontent.com/5889404/157501597-85215b86-be4a-439e-bf06-9e046df72eda.png">

`Reference` need to be serilaized to fix dataframe codegen issue with Immutables.
`Content` need to be  serilaized for storing in spark row. 
So, this PR has complete changes from #3561. If we merge it, I can rebase it.

You can take a look this PR, There is one more work pending to flatten (aka explode) the `expired_contents` column.
If I do that, Then I don't have to keep `IdentifiedResult` class (it was a map for keeping all expired contents under content id). Test framework depends on 'IdentifiedResult'. So, cleaning it will bloat the PR. Hence I am planning to do in a follow up.
But we still need to show  `Content` in the spark output row. so #3561 is needed. 

Please tell me your opinion on this.

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifyContentsPerExecutor.java:
```diff
@@ -62,9 +67,34 @@ public IdentifyContentsPerExecutor(GCParams gcParams) {
+    List<Row> rows = new ArrayList<>();
+    references.foreach(
+        reference -> {
+          rows.addAll(computeExpiredContents(liveContentsBloomFilterMap, reference));
```
TODO: Accumulating all the rows here will bloat up the driver memory. Need to pushdown this iterator to Nessie commit log traversal.

ajantha-bhat replied:
Finished the TODO.

ajantha-bhat commented:
@snazy , @rymurr :
![image](https://user-images.githubusercontent.com/5889404/157724141-3db9b058-381f-4b14-9df1-ad7f38649a3f.png)

The sample output looks like above. 
**I have removed `IdentifiedResults`, `ContentValues` (intermediate classes) completely and fixed test framework to use dataframe output directly. 
Also removed `Serializable` for `Contents`**

There is one TODO: I need to push Iterator<Row> of row to commit log level to properly and merge the iterator from each task instead of collect the rows. I will explore it tomorrow.

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifyContentsPerExecutor.java:
```diff
@@ -267,4 +204,120 @@ private Instant getCutoffTimeForRef(String reference, Map<String, Instant> dropp
+    // Worst case few expired contents will be considered live due to bloom filter
+    // fpp.
+    // But live contents never be considered as expired.
+    Predicate<Content> getExpiredContent =
```
"get" is misleading here - either no verb or "filter" or "testFor"

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java:
```diff
@@ -0,0 +1,269 @@
+  private static final String COL_CONTENT_ID = "contentId";
+  private static final String COL_CONTENT_TYPE = "contentType";
+  private static final String COL_SNAPSHOT_ID = "snapshotId";
+  private static final String COL_METADATA_LOCATION = "metadataLocation";
```
Why is "metadataLocation" needed?

ajantha-bhat replied:
If the content is dropped from all the reference (may not be always), I can just use `metadataLocation` and call Iceberg delete_reachable_files action. 

basically this info is for the consumer of GC results, similar to snapshot id.

ajantha-bhat commented on servers/jax-rs-tests/pom.xml:
```diff
@@ -99,6 +99,12 @@
     </dependency>
+    <dependency>
+      <groupId>org.projectnessie</groupId>
+      <artifactId>nessie-spark-extensions</artifactId>
```
Because usually the Nessie GC is used with Nessie + spark + iceberg.
Nessie only works with spark3.1 with Iceberg[1], so used the same environment settings here.

[1]As per compatibility doc
https://github.com/projectnessie/nessie#compatibility

ajantha-bhat commented on servers/jax-rs-tests/pom.xml:
```diff
@@ -124,6 +130,11 @@
     </dependency>
+    <dependency>
+      <groupId>org.apache.iceberg</groupId>
+      <artifactId>iceberg-spark3-runtime</artifactId>
```
Totally agree. But now it is not failing because there is conflict. But it can fail once we upgrade the versions.
As it a whole testcase refactor, I have to figure out / adjust this testcases. **So, will do it in an immediate follow up** if it is ok. 

 #3601 - Created this ticket to track it. 

I am planning to move these tests to the module (`nessie-spark-extensions-base`) that already depends on
`iceberg-spark3-runtime`. But it will become IT test. which should be fine as it is anyway end to end test.

ajantha-bhat replied:
Also thought of adding individual jars and exclude nessie-client from iceberg. But once Iceberg upgrades these testcases will fail in that case. So, I felt moving to `nessie-spark-extensions-base` is the right way.

ajantha-bhat replied:
added exclusions to the iceberg's nessie classes and depending on the latest nessie version. 

with a follow up of https://github.com/projectnessie/nessie/issues/3692 once Iceberg makes a release.

ajantha-bhat commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGC.java:
```diff
@@ -19,39 +19,44 @@
 
 public abstract class AbstractRestGC extends AbstractRest {
 
+  @TempDir static File LOCAL_DIR;
```
yes, this warehouse is used only for GC output table. but as the table name was same for all the testcase, it was using the same location. I will use different table names for each testcase.

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifyContentsPerExecutor.java:
```diff
@@ -267,4 +204,120 @@ private Instant getCutoffTimeForRef(String reference, Map<String, Instant> dropp
+                  OptionalInt.empty())
+              .filter(getUnprotectedCommits)
+              .map(LogResponse.LogEntry::getOperations)
+              .filter(operation -> operation instanceof Operation.Put)
```
I wanted to filter out only the PUT operation. As the `operation` is not having any type filed. This was the only way I could find. 
Any suggestion?

`operation instanceof Operation.Put` return true for `PUT` and false for `DELETE` operation.

ajantha-bhat replied:
or you meant use this ? `.filter(Operation.Put.class::isInstance)`

ajantha-bhat commented:
@snazy: Addressed the comments other than Uber jar one. Created the #3601 follow up for fixing that.

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java:
```diff
@@ -0,0 +1,270 @@
+  private static final String COL_CONTENT_ID = "contentId";
+  private static final String COL_CONTENT_TYPE = "contentType";
+  private static final String COL_SNAPSHOT_ID = "snapshotId";
+  private static final String COL_METADATA_LOCATION = "metadataLocation";
```
It's still wrong to even consider this value.
Nothing guarantees that the stored value will be valid wehn GC actually runs.

ajantha-bhat replied:
oh. yeah global state changes.

ajantha-bhat replied:
removed now.
 
Actually first time comment was just 
> Why is "metadataLocation" needed?

So, I replied why did I used it.  So, please provide a bit more context/expected result while working with me to avoid rework or re-review. Thanks.

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifyContentsPerExecutor.java:
```diff
@@ -267,4 +204,120 @@ private Instant getCutoffTimeForRef(String reference, Map<String, Instant> dropp
+                  OptionalInt.empty())
+              .filter(unprotectedCommitsPredicate)
+              .map(LogResponse.LogEntry::getOperations)
+              .filter(operation -> operation instanceof Operation.Put)
```
This is still wrong.

ajantha-bhat replied:
Then how to filter the `PUT` operation ? can you please suggest? 
Do I need to implement a `operation.type` ?

ajantha-bhat replied:
or you meant use this ? .filter(Operation.Put.class::isInstance)

ajantha-bhat replied:
pushed changes now by using `Operation.Put.class::isInstance`

ajantha-bhat replied:
Oh, it is a list of operations, so checking it against instance is wrong. I get it now !
Also figured out that test case didn't catch because validation is wrong.

It should be `dfExpected.except(dfActual)` instead of `dfActual.except(dfExpected)` because if the actual is empty, except() can still return empty.

snazy replied:
> It should be dfExpected.except(dfActual) instead of dfActual.except(dfExpected) because if the actual is empty, except() can still return empty.

Where's that??

ajantha-bhat replied:
oh, In my code it is here:
`AbstractRestGC#verify()`

snazy reviewed (changes requested):
> Addressed the comments other than Uber jar one.

No, you haven't addressed all comments.

ajantha-bhat commented:
Cerated [#3603](https://github.com/projectnessie/nessie/pull/3603) to handle uber jar (#3601) issue. Once it is merged, I can rebase this PR.

ajantha-bhat commented:
> @snazy: Addressed the comments other than Uber jar one. Created the https://github.com/projectnessie/nessie/issues/3601 follow up for fixing that.

Fixed the Uber jar impact by moving test cases to GC module in a separate PR and rebased this PR. 

I can still restrict to use `${client.nessie.version}`in gc-base module but Iceberg_view and few other changes are not in `${client.nessie.version}` (aka 0.18.0), So, removing them and adding them is rework. So, Instead of that as there is no breaking changes, I am using the project version itself. 
Once the next Iceberg release is done, I will change it to project version.

ajantha-bhat commented:
ping... @snazy

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifyContentsPerExecutor.java:
```diff
@@ -267,4 +204,118 @@ private Instant getCutoffTimeForRef(String reference, Map<String, Instant> dropp
+        });
+    // merge the iterators form each task.
+    Iterator<Row> mergedIterator = Iterators.concat(iterators.iterator());
+    return scala.collection.JavaConverters.asScalaIterator(mergedIterator);
```
Add import for this

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java:
```diff
@@ -0,0 +1,261 @@
+   * @param runId run id of current identify task.
+   * @param startedAt task start time.
+   */
+  void writeToOutputTable(Dataset<Row> rowDataset, String runId, Timestamp startedAt) {
```
This function writes the results of all tasks and then writes the marker row. Why does it need the marker row then?

Also: why don't the executors write the rows directly but instead pass the rows back to the caller?

So either write the rows from the executors plus the marker row here or don't use the marker row at all.

What are the time + space requirements (the overall algorithm) for both approaches wrt to writing to the output table? Have you tested this approach with 10000 or 100000 tables?

ajantha-bhat replied:
> Also: why don't the executors write the rows directly but instead pass the rows back to the caller?

Executor will not have spark context to create the dataframe[1]. Hence, driver collects the row iteratively (not all at once) and writes it to the output table. However overall operation is atomic in nature and no need of marker row. I will remove it.
[1] https://stackoverflow.com/questions/48715661/spark-how-can-i-create-local-dataframe-in-each-executor

> What are the time + space requirements (the overall algorithm) for both approaches wrt to writing to the output table? Have you tested this approach with 10000 or 100000 tables?

It depends on the parallelism, available cores and other configurations. Number of references will be divided into X (available parallelism) and distributively each reference will reverse its commit log. As there is no alternative approach available at the moment, I didn't test performance. Want to have a fully functional working algorithm first and take up the performance optimization in a follow up.

ajantha-bhat replied:
say If the number of executor instance is configured as 5,
there will be 5 task launched with an iterator of 2000 references per each task.
5 tasks execute concurrently in finding the expired row per task (that too one task will produce one expired row at a time form Row iterator by working on one reference)
These 5 rows are added to in memory and again next batch of 5 rows are fetched and so on...
As the final action is `.append()`, once the default target file is reached (say 512mb), rows will be flushed to parquet file.

Based on the cpu and memory available, we can tune the executor instance.

Do you have any other suggestion for this?

snazy commented on gc/gc-base/src/test/java/org/projectnessie/gc/base/AbstractRestGCRepoTest.java:
```diff
@@ -0,0 +1,194 @@
+public abstract class AbstractRestGCRepoTest extends AbstractRestGCTest {
+
+  private final String catalogName = "nessie";
+  private final String identifierTableName = "identified_results";
```
identifier? name?

snazy commented on gc/gc-base/src/test/java/org/projectnessie/gc/base/AbstractRestGC.java:
```diff
@@ -19,35 +19,41 @@
 
 public abstract class AbstractRestGC extends AbstractRest {
 
+  @TempDir static File LOCAL_DIR;
```
This (`static`) leaks data from tests leak into other tests, which is not a good practice.

snazy commented on gc/gc-base/src/test/java/org/projectnessie/gc/base/AbstractRestGCRepoTest.java:
```diff
@@ -0,0 +1,194 @@
+    getOrCreateEmptyReference(getApi(), gcRefName);
+    SparkSession sparkSession = null;
+    try {
+      sparkSession = getSparkSession();
```
try-with-resources

snazy commented on gc/gc-base/src/test/java/org/projectnessie/gc/base/AbstractRestGCRepoTest.java:
```diff
@@ -0,0 +1,194 @@
+    getOrCreateEmptyReference(getApi(), gcRefName);
+    SparkSession sparkSession = null;
+    try {
+      sparkSession = getSparkSession();
```
try-with-resources

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java:
```diff
@@ -0,0 +1,261 @@
+                .sparkContext()
+                .conf()
+                .getAllWithPrefix(String.format("spark.sql.catalog.%s.", catalog)));
+    if (branch != null) {
```
When can branch be null?

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java:
```diff
@@ -0,0 +1,261 @@
+      String contentId, String contentType, long snapshotId, String refName, String hash) {
+    return RowFactory.create(
+        IdentifiedResultsRepo.TYPE_CONTENT,
+        // the fixed value columns (runId and startTime)
```
Why not just pass those values downstream instead this apporach?

snazy replied:
Really better create the `Row` as it will look like instead of manipulating it later.

ajantha-bhat replied:
No problem, I will change it.

ajantha-bhat replied:
fixed.

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java:
```diff
@@ -0,0 +1,261 @@
+            catalogConfWithRef(sparkSession, catalogName, gcRefName),
+            sparkSession.sparkContext().hadoopConfiguration());
+    try {
+      nessieCatalog.createTable(tableIdentifier, icebergSchema);
```
How do you prevent that GC doesn't expire it's own data?

ajantha-bhat replied:
I think based on previous internal discussion.
GC output table should be considered as another normal iceberg table.
So it should be OK to GC its own contents and expire them. When I add a documentation on GC, I will update this part.

snazy replied:
Documentation isn't enough IMO.
There will be users that will shoot themselves into the foot.

ajantha-bhat replied:
GC output table is just another iceberg table created in separate reference.

> The fact that GC can wipe its own data

I am not sure about above statement, 

a) If the gc output table is not dropped,
gc don't delete the actual gc output rows/data, only redundant expired metadata (like snapshot id form metadata) will be removed.
b) If the gc output table is dropped, as it is another Nessie managed Iceberg table, actual data of dropped gc output table will be cleaned from GC.

Are you suggesting don't keep it as Nessie managed table? or skip gc for this table?

tomekl007 replied:
> or skip gc for this table?
How many dates the GC will write? Wouldn't it lead to too much data being kept? Maybe we should have a similar `cutoffPeriod` for the GC table as well? (but it would be substantially higher)

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/GCParams.java:
```diff
@@ -70,6 +70,20 @@ default double getBloomFilterFpp() {
+   */
+  String getOutputTableRefName();
+
+  /** Output table name to be used for storing the results in {@link #getOutputTableRefName()}. */
```
ok. It should be table identifier (namespace + table name)

ajantha-bhat commented:
@snazy : Pushed the changes in the separate commit, replied to the unhandled comments/questions. You can reply to the unhandled comments if anything needs to handled or my understanding is wrong. Thanks.

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java:
```diff
@@ -0,0 +1,196 @@
+    //    (SELECT MAX(gcRunStart) FROM nessie.db2.`identified_results@someGcBranch`) LIMIT 1
+    List<Row> rows =
+        sql(
+                "SELECT %s FROM %s WHERE %s = (SELECT MAX(%s) FROM %s) LIMIT 1",
```
`DISTINCT` missing

snazy commented:
Still two major concerns:
* The amount of data handled in `identifyExpiredContents`
* The fact that GC can wipe its own data.

ajantha-bhat commented:
>Still two major concerns:
The amount of data handled in identifyExpiredContents
The fact that GC can wipe its own data.

I have replied my thoughts on the same. Also removed the late filling of run id and gc start time.

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifyContentsPerExecutor.java:
```diff
@@ -267,4 +209,127 @@ private Instant getCutoffTimeForRef(String reference, Map<String, Instant> dropp
   }
+
+  // --- Methods for computing expired content rows ----------
+  /**
```
> Mixtrure of Java and Scala iterators

I might have to write `DistributedIdentifyContents` and `IdentifyContentsPerExecutor` in scala :(

> Check and fix usage and lifetime of NessieApi

Yes, In driver I can close the NessieApi before calling `distributedIdentifyContents.identifyExpiredContents` in GCImpl.
Executor creates is creating NessieApi per reference, I can make it per task. 

>Ensure that only one Row for each content/snapshot-ID is returned

Currently rows are returned per reference, say if the snapshot id 42 is expired for contentId_1 in 5 different references, there will be 5 similar rows for contentID-snapshotID (contentId_1 - 42) with different ref@hash in the total output.

I can add distinct here, but we lose reference info if I do that, also adding aggregate function require computation of whole data. So, it will bloat up.
In the consumer (while querying the output table), we can easily get the distinct values.

tomekl007 replied:
>Ensure that only one Row for each content/snapshot-ID is returned
Maybe you need to remodel the structure to capture the fact that you have a list of smth per the `content/snapshot-ID`?

ajantha-bhat replied:
> Maybe you need to remodel the structure to capture the fact that you have a list of smth per the content/snapshot-ID?

That needs data shuffle (aggregate function) and needs entire data. It will bloat up. Hence, handing in the query is better as it will have filter + aggregate.

ajantha-bhat commented:
> Mixtrure of Java and Scala iterators

Underlaying Nessie API has java stream / iterator.
Hence, java to scala iterator conversion is required even if I write all these code in scala.

> Check and fix usage and lifetime of NessieApi

Fixed.

> Ensure that only one Row for each content/snapshot-ID is returned

Currently rows are returned per reference, say if the snapshot id 42 is expired for contentId_1 in 5 different references, there will be 5 similar rows for contentID-snapshotID (contentId_1 - 42) with different ref@hash in the total output.

I can add DISTINCT here while writing, but reference info will be lost if I do that, also adding aggregate function require computation of whole data. So, it will bloat up.
In the consumer (while querying the output table), we can easily get the distinct values. Hence, better to handle at consumer side.

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifyContentsPerExecutor.java:
```diff
@@ -267,4 +209,119 @@ private Instant getCutoffTimeForRef(String reference, Map<String, Instant> dropp
+          });
+    }
+    // merge the iterators from each task.
+    Iterator<Row> mergedIterator = Iterators.concat(iterators.iterator());
```
You're still using/returning _*closed*_ `NessieApi` instances here!!

snazy reviewed (changes requested):
Still open questions:
* How do you 100% ensure that GC does _not_ GC it's own data?
* Can you 100% ensure that there will be no OutOfMemoryError ever even when using 10000 references and 100000 contents per reference?

ajantha-bhat commented:
> How do you 100% ensure that GC does not GC it's own data?
while running gc, because GC table is not dropped, only old snapshots from gc output table will be removed from metadata from expire_snapshots.
I can add a validation to query GC table before and after GC to show that data is not deleted (once expiry procedure is added in the further PR)

> Can you 100% ensure that there will be no OutOfMemoryError ever even when using 10000 references and 100000 contents per reference?
I think it depends on spark configuration available. At least default file size 512 MB of driver memory is required.

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifyContentsPerExecutor.java:
```diff
@@ -267,4 +209,119 @@ private Instant getCutoffTimeForRef(String reference, Map<String, Instant> dropp
+            return reference;
+          });
+      // merge the iterators from each task.
+      Iterator<Row> mergedIterator = Iterators.concat(iterators.iterator());
```
Please think again - this is *NOT* fixing the issue!

ajantha-bhat replied:
The test case was passing, let me understand how it is working.

tomekl007 replied:
Maybe we can convert the iterator to DF?

ajantha-bhat replied:
> Maybe we can convert the iterator to DF?

this is in executor and spark context is not available in executor. map partitions need the iterator<ROW> as output

ajantha-bhat replied:
I can close it when the iterator has no more rows per reference. 
For this instead of one API per task, now I have to create one API per reference.

snazy replied:
This is just moving the problem - not fixing it.

Now you create one API instance per reference, which is just not necessary.
Further, there are cases when the API instance is not closed - for example if not all items have been fetched from the iterator. I.e. you're introducing resource leaks.

Maybe look at how external resources are managed in Spark?

ajantha-bhat replied:
> Maybe look at how external resources are managed in Spark?

a) I tried checking it yesterday. Spark suggests broadcasting the resource and cleaning it in driver. As mentioned below in other comment, some things are not serailizable in `NessieHttpClient`. Hence it is a problem.

b) I used this solution in the current implementation
https://stackoverflow.com/questions/36545579/spark-how-to-use-mappartition-and-create-close-connection-per-partition

c) `Now you create one API instance per reference, which is just not necessary.`
because iterators are merged, I cannot check hasNext() of final iterator as it happens in spark code gen. Hence , moved to per reference.

d)`for example if not all items have been fetched from the iterator`, 
this can only happen if task is aborted by the user or executor killed (if executor killed, no need to worry about resource leak on that executor as it is mapped to one JVM). If the task is aborted, I don't have a way to close the api in this situation :(

snazy replied:
I don't believe that there is no way to properly fix it.
The "not serializable" argument is not valid IMO - otherwise JDBC would not work, CSV file access would not work, etc etc.

tomekl007 replied:
@ajantha-bhat you should look for ways to share state per executor, maybe: https://able.bio/patrickcording/sharing-objects-in-spark--58x4gbf
will be helpful?
So you should basically create a static `NessieAPI`(that will be initialized per executor)

ajantha-bhat replied:
@tomekl007 : Even If I make a static instance at executor, when to close() is till a problem. ( I wouldn't know when all the iterators are consumed, I will only know when each iterator is consumed)

In current code, I am closing when the iterator is consumed fully or any exception in filling iterator. 

But as @snazy said, what if iterator is not consumed fully due to some error (as the final action is `.append()`, iterator is consumed by spark for append() operation, if there is some failure) I don't have a way to close it.

I might have to explore towards serializing the API, but that involves serializing a ton of classes along with it.

ajantha-bhat replied:
@snazy : As people are not happy with #3793 (including me). I don't think I have a solution for this other than consuming the iterator and cleaning the resource as mentioned above. But that solution will bloat up the executor.

So, At the moment I don't have any solution.

ajantha-bhat replied:
Fixed. Thanks.

tomekl007 commented on gc/gc-base/src/test/java/org/projectnessie/gc/base/AbstractRestGC.java:
```diff
@@ -99,93 +113,56 @@ void performGc(
-    }
+  private void verify(
+      Dataset<Row> actual, List<Row> expectedRows, SparkSession session, StructType schema) {
+    Dataset<Row> expected = session.createDataFrame(expectedRows, schema);
```
Global note regarding scaling/paralellization. 
I suspect that most often a table has only one ref: `main`. 
In our model, the ref == partition. In that scenario, when we run the job for one table we won't be able to achieve any parallelization. Only one executor will be processing the only ref (`main`) for this table.
@ajantha-bhat Did we think about it?

ajantha-bhat replied:
I thought about it before, but I cannot distributively traverse commits. It is currently paginated sequential traversal. 
So, I cannot have a commit level distribution instead of reference level.

May be we need to modify the backend to fetch commit log distributively in future.

tomekl007 replied:
so currently, in most cases (only `main` branch) the processing won't be distributed at all, isn't it?

ajantha-bhat replied:
> so currently, in most cases (only main branch) the processing won't be distributed at all, isn't it?

yes.

But not sure if only main branch is used in most cases. One of the main use case of nessie is branching and tagging.
Main branch can have more commits compared to other branches. But backend doesn't support distributed query of commit logs as of now.

tomekl007 commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/DistributedIdentifyContents.java:
```diff
@@ -65,21 +71,27 @@ public DistributedIdentifyContents(SparkSession session, GCParams gcParams) {
       Map<String, ContentBloomFilter> liveContentsBloomFilterMap, List<String> references) {
-
+    String runId = UUID.randomUUID().toString();
+    Timestamp startedAt = Timestamp.from(Instant.now());
```
nit: allow injection of `Instant.now()` for testing

ajantha-bhat replied:
over all (Instant like commit time, author time etc) in GC cannot be mocked as Nessie backend runs in quarkus server and long back faced many problems to mock the time in quarkus server. 

Plus this is not an input argument, this becomes output argument. Mocking is not necessary.

tomekl007 commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java:
```diff
@@ -0,0 +1,202 @@
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(IdentifiedResultsRepo.class);
+
+  private static final String COL_GC_RUN_START = "gcRunStart";
```
Column names should be at least package-private, and we should write an integration test for fetching those from the target table.

ajantha-bhat replied:
there is a public interface `getSchema()`, which contains the column names.
So, I think no need of column name to be package private.
already testcases exist that use `schema` to create expected rows.

tomekl007 commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java:
```diff
@@ -0,0 +1,202 @@
+    // SELECT * FROM nessie.db1.`identified_results@someGcRef`
+    //    WHERE gcRunId = '9e809dc9-ac35-41c9-b1c0-5ebcf452ea2d'
+    return sql(
+        "SELECT * FROM %s WHERE %s = '%s'",
```
do we need to do `SELECT *`? Shouldn't we pass column names explicitly and fetch only what's needed?

tomekl007 replied:
yeah, but in case a new column is added to this table, we will fetch it and maybe not necessarily

ajantha-bhat replied:
This table is purely for GC consumer, if new column is added then it is required for the consumer. 
I will reevaluate this again while adding GC consumer code.

tomekl007 commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java:
```diff
@@ -0,0 +1,202 @@
+                COL_GC_RUN_START,
+                catalogAndTableWithRefName)
+            .collectAsList();
+    return rows.isEmpty() ? null : rows.get(0).getString(0);
```
maybe return `Optional`?

tomekl007 commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java:
```diff
@@ -0,0 +1,202 @@
+      // write content rows to the output table
+      rowDataset.writeTo(catalogAndTableWithRefName).append();
+    } catch (NoSuchTableException e) {
+      throw new RuntimeException(e);
```
can you give a descriptive message? smth like: `"Problem when writing the rowDataset: " +  rowDataset + " to " + catalogAndTableWithRefName`

tomekl007 commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java:
```diff
@@ -0,0 +1,202 @@
+    try {
+      nessieCatalog.createTable(tableIdentifier, icebergSchema);
+    } catch (AlreadyExistsException ex) {
+      // Table can exist from previous GC run, no need to throw exception.
```
isn't there an API in Nessie that allows us to specify whether the table should be created only when missing? Instead of throwing an exception as a way to signal the logic? If not, maybe we should consider adding such a method in nessie?

ajantha-bhat replied:
This is Iceberg API to create a table. IIRC Only this was available.

tomekl007 commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java:
```diff
@@ -0,0 +1,202 @@
+        + "."
+        + namespace
+        + "."
+        + ImmutableTableReference.builder().name(tableName).reference(refName).build();
```
so we are relying on the auto-generated `toString` method for the `ImmutableTableReference`. Is it safe?

ajantha-bhat replied:
`TableReference` has a `toString() `override. Immutables will use that.

tomekl007 commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java:
```diff
@@ -0,0 +1,202 @@
+                .sparkContext()
+                .conf()
+                .getAllWithPrefix(String.format("spark.sql.catalog.%s.", catalog)));
+    return conf.map(t -> t._1.equals("ref") ? Tuple2.apply(t._1, branch) : t)
```
so your purpose to filter the `ref` and map it to `branch`, right? Maybe this logic would be more descriptive when using `filter`?

ajantha-bhat replied:
> so your purpose to filter the ref and map it to branch, right?

yes.

tomekl007 replied:
can you add a comment explaining that to this method

ajantha-bhat replied:
added.

tomekl007 commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifyContentsPerExecutor.java:
```diff
@@ -62,11 +71,15 @@ public IdentifyContentsPerExecutor(GCParams gcParams) {
-  protected Function<String, IdentifiedResult> computeExpiredContentsFunc(
-      Map<String, ContentBloomFilter> liveContentsBloomFilterMap) {
-    return reference -> computeExpiredContents(liveContentsBloomFilterMap, reference);
+  protected SerializableFunction1<scala.collection.Iterator<String>, scala.collection.Iterator<Row>>
```
Is it a bit weird that we mix `SerializableFunction1` with java code? Maybe you can make this method static and prevent the need for serialization? (and use `Function`)

ajantha-bhat replied:
I tired having all static methods, still gets error for  `interface scala.Function1`
That is why I previously had `SerializableFunction1`

```
Caused by: java.io.NotSerializableException: org.projectnessie.gc.base.IdentifyContentsPerExecutor$$Lambda$2943/0x000000080147d040
Serialization stack:
	- object not serializable (class: org.projectnessie.gc.base.IdentifyContentsPerExecutor$$Lambda$2943/0x000000080147d040, value: org.projectnessie.gc.base.IdentifyContentsPerExecutor$$Lambda$2943/0x000000080147d040@359ae83a)
	- field (class: org.apache.spark.sql.execution.MapPartitionsExec, name: func, type: interface scala.Function1)
```

My branch with this change, if you would like to play around

https://github.com/ajantha-bhat/nessie/commit/aa43d4e72a29a95023e9082de4f57085eb93fd3d

tomekl007 commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifyContentsPerExecutor.java:
```diff
@@ -267,4 +209,119 @@ private Instant getCutoffTimeForRef(String reference, Map<String, Instant> dropp
+   * <p>Each reference can produce one or more rows.
+   */
+  private scala.collection.Iterator<Row> getExpiredContentRows(
+      scala.collection.Iterator<String> references,
```
Maybe we can convert the iterator to DF and use a Spark API for transformation, instead of using the `scala.collection.Iterator`?

ajantha-bhat replied:
this is in executor and spark context is not available in executor. map partitions need the iterator as output

tomekl007 commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/SerializableFunction1.java:
```diff
@@ -0,0 +1,21 @@
+import java.io.Serializable;
+
+/** Interface that makes scala.Function1 as Serializable. */
+public interface SerializableFunction1<T, U> extends scala.Function1<T, U>, Serializable {}
```
can you try using static methods instead of passing this function?

ajantha-bhat replied:
tried, replied above to similar coment.

tomekl007 commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifyContentsPerExecutor.java:
```diff
@@ -267,4 +209,119 @@ private Instant getCutoffTimeForRef(String reference, Map<String, Instant> dropp
+      String runId,
+      Timestamp startedAt) {
+    List<Iterator<Row>> iterators = new ArrayList<>();
+    try (NessieApiV1 api = GCUtil.getApi(gcParams.getNessieClientConfigs())) {
```
You shouldn't use the try-with-resources.
The spark transformations are lazy and async, and the try-with-resources won't know when it should close the Nessie API.

you can try something like this:
On the driver, create a `Broadcast` variable of `NessieApiV1`
It is a way of passing states in Sparks such as clients between driver and executors.
You should not close it

Then pass the `Broadcast<NessieAPI>` into your transformation
it will create N instances of nessieAPI, one per executor.
After all, the processing is done (and after saving to a table) you need to close the Nessie API explicitly.

ajantha-bhat replied:
how it brodcasts? I assuming by serilaizing it and some inside content of `NessieApiV1` has Http stuff, which are not serializable.

> Alternatively, you can keep creating API on the executor as is now, but you need to take care of closing it manually once all the processing is done.

I just return iterator of rows to driver and so spark it self will process the iterator, call doesn't comes back to executor to close it.

ajantha-bhat replied:
Tried broadcast, broadcast goes for serialization.

I had to make 10 plus classes serializable for http client, even after that `com.fasterxml.jackson.databind.jsontype.impl.TypeNameIdResolver` is not serializable

ajantha-bhat commented:
@snazy, @tomekl007 : Fixed the api close issue by using spark's TaskCompletionListener and also made the api instance per task instead of per reference. Thanks.

Please have a look at it again.

tomekl007 commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifyContentsPerExecutor.java:
```diff
@@ -62,45 +75,46 @@ public IdentifyContentsPerExecutor(GCParams gcParams) {
-
-      return walkLiveCommitsInReference(gcStateParamsPerTask);
+    NessieApiV1 api = GCUtil.getApi(gcParams.getNessieClientConfigs());
+    TaskContext.get()
```
nice 👍

tomekl007 reviewed (dismissed):
LGTM, thank you for taking the time for investigating the close behavior of NessieAPI and implementing the `close()` in the listener here:
https://github.com/projectnessie/nessie/pull/3421/files#diff-1bf51914a6143465d680b79d8ebcec9486ff4093c10dd712a6d443ff1fb48583R91

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifyContentsPerExecutor.java:
```diff
@@ -267,4 +218,124 @@ private Instant getCutoffTimeForRef(String reference, Map<String, Instant> dropp
+              LOGGER.info("Closing the nessie api for compute expired contents task");
+              api.close();
+            });
+    List<Iterator<Row>> iterators = new ArrayList<>();
```
This can all be replaced with this one - no need to collect + re-stream.
```
    return references.flatMap(
        reference ->
            JavaConverters.asScalaIterator(
                    walkAllCommitsInReference(
                        api,
                        GCUtil.deserializeReference(reference),
                        liveContentsBloomFilterMap,
                        runId,
                        startedAt))
                .toTraversable());
```

ajantha-bhat replied:
thanks. Replaced.

snazy reviewed (changes requested):
Need to simplify `getExpiredContentRows` and then it should be good.
